// stimulus-parser/dist/controller_definition.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ControllerDefinition = exports.defaultValuesForType = void 0;
const stimulus_webpack_helpers_1 = require("@hotwired/stimulus-webpack-helpers");
exports.defaultValuesForType = {
    Array: [],
    Boolean: false,
    Number: 0,
    Object: {},
    String: "",
};
class ControllerDefinition {
    static controllerPathForIdentifier(identifier) {
        const path = identifier.replace(/--/g, "/").replace(/-/g, "_");
        return `${path}_controller.js`;
    }
    constructor(project, path) {
        this.methods = [];
        this.targets = [];
        this.classes = [];
        this.values = {};
        this.project = project;
        this.path = path;
    }
    get controllerPath() {
        return this.project.relativeControllerPath(this.path);
    }
    get identifier() {
        return (0, stimulus_webpack_helpers_1.identifierForContextKey)(this.controllerPath) || "";
    }
    get isNamespaced() {
        return this.identifier.includes("--");
    }
    get namespace() {
        const splits = this.identifier.split("--");
        return splits.slice(0, splits.length - 1).join("--");
    }
}
exports.ControllerDefinition = ControllerDefinition;
